### PR TITLE
Feature/lji 6 employe mvc

### PR DESCRIPTION
**Pull request by Cioperis**
Added employee mvc template
Added some methods to add an employee, update, delete, get all employees.
Data validation to be added and tests

### CODE REVIEW

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class EmployeeService {
+    private final EmployeeRepository employeeRepository;
+    private final EmployeeMapper employeeMapper;
+
+    public EmployeeDTO addEmployee(EmployeeDTO employeeDTO){
+        return EmployeeDTO.builder()
```
[the remark this answers is not in the document]
Comment (anton-kolyaev) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:19]: You've added mappers to convert dto to entity and back. Do you still need in manual remapping?

File: src/main/resources/changelog/changelog-master.xml [hunk 2]
```diff
@@ -1,11 +1,13 @@
 <?xml version="1.0" encoding="UTF-8"?>
 
 <databaseChangeLog
-  xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
-  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
```
Comment (anton-kolyaev) [inline src/main/resources/changelog/changelog-master.xml:4]: avoid changes like this (auto format code) on common files - if it's necessary discuss and adjust codestyle for everyone in team

**Review by anton-kolyaev — CHANGES_REQUESTED** [review]

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                            .phoneNumber(employee.getPhoneNumber())
+                            .build();
+                })
+                .orElse(null);
+    }
+
+    public void deleteEmployee(UUID employeeId){
+        employeeRepository.deleteById(employeeId);
+    }
+}
```
Comment (anddevdev) [inline src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:82]: I think we should use soft-delete

File: src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import com.coherentsolutions.pot.insurance.dto.EmployeeDTO;
+import com.coherentsolutions.pot.insurance.service.EmployeeService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/v1/employee")
```
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java]: fixed

File: src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+@Builder
+public class EmployeeEntity {
+    @Id
+    private UUID employeeId;
+
+    private String firstName;
+
+    private String lastName;
+
+    private String userName; // unique field
```
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java:28]: fixed

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                .map(employee -> EmployeeDTO.builder()
+                        .employeeId(employee.getEmployeeId())
+                        .firstName(employee.getFirstName())
+                        .lastName(employee.getLastName())
+                        .userName(employee.getUserName())
+                        .dateOfBirth(String.valueOf(employee.getDateOfBirth()))
+                        .SSN(employee.getSSN())
+                        .phoneNumber(employee.getPhoneNumber())
+                        .build())
+                .orElse(null);
```
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: resolved but i used responsestatusexception here, is that sufficient or should i create and use a custom exception?

File: src/main/resources/changelog/changelog-master.xml (hunk 2 above)
Comment (Cioperis) [reply src/main/resources/changelog/changelog-master.xml]: reverted and added the closing tab </databaseChangeLog> cause it was missing for some reason

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: added proper mapping (not manual) except to update information

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 3 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: fixed

File: src/main/java/com/coherentsolutions/pot/insurance/dto/EmployeeDTO.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+public class EmployeeDTO {
+    private UUID employeeId;
+    private String firstName;
+    private String lastName;
+    private String userName;
+    private String email;
+    private LocalDate dateOfBirth;
+    private Integer SSN;
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/dto/EmployeeDTO.java:21]: lowercase property name

File: src/main/java/com/coherentsolutions/pot/insurance/repository/EmployeeRepository.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.coherentsolutions.pot.insurance.repository;
+
+import com.coherentsolutions.pot.insurance.entity.EmployeeEntity;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.UUID;
+
+public interface EmployeeRepository extends JpaRepository<EmployeeEntity, UUID>{
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/repository/EmployeeRepository.java:8]: @Repository annotation

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+
+import java.util.List;
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class EmployeeService {
+    private final EmployeeRepository employeeRepository;
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:17]: There is no need to inject a mapper, you already have INSTANCe of the mapper

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    private final EmployeeMapper employeeMapper;
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:18]: There is no need to inject a mapper, you already have INSTANCe of the mapper

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    public List<EmployeeDTO> getAllEmployees(){
+        return employeeRepository.findAll().stream()
+                .map(employeeMapper::employeeToEmployeeDTO)
+                .toList();
+    }
+
+    public EmployeeDTO updateEmployee(UUID employeeId, EmployeeDTO updatedEmployeeDTO){
+        return employeeRepository.findById(employeeId)
+                .map(employee -> {
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:39]: Use the mapper to update the entity and avoid doing it manually

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                    employee.setPhoneNumber(updatedEmployeeDTO.getPhoneNumber());
+                    return employeeMapper.employeeToEmployeeDTO(employeeRepository.save(employee));
+                })
+                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Error 404: Employee not found with id: " + employeeId));
+    }
+
+    public void deleteEmployee(UUID employeeId){
+        employeeRepository.findById(employeeId)
+                .map(employee -> {
+                    employee.setDeleted(true);
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:54]: Do we need to have such property or can we add status for employees and set it to "INACTIVE"?

File: src/main/resources/changelog/changelog-master.xml
Comment (SavaLX) [inline src/main/resources/changelog/changelog-master.xml:1]: Where is your changeset with creating the Employee table and inserting some data into the Database?

File: src/main/java/com/coherentsolutions/pot/insurance/dto/EmployeeDTO.java (hunk 7 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/dto/EmployeeDTO.java]: fixed

File: src/main/java/com/coherentsolutions/pot/insurance/repository/EmployeeRepository.java (hunk 8 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/repository/EmployeeRepository.java:10]: fixed

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 10 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: instance is now used

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 11 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:38]: manual mapping removed

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: changed to reuse existing exceptions

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 12 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: updated to use enum inactive status instead

File: src/main/resources/changelog/changelog-master.xml
Comment (Cioperis) [reply src/main/resources/changelog/changelog-master.xml:1]: resolved

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: update information was changed too to not use manual remapping

File: src/test/java/com/coherentsolutions/pot/insurance/claim/EmployeeControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+import java.util.List;
+import java.util.UUID;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
+class EmployeeControllerTest {
```
Comment (anton-kolyaev) [inline src/test/java/com/coherentsolutions/pot/insurance/claim/EmployeeControllerTest.java:21]: Use mockMVC to implement integration tests and lets write unit tests for something meaningful

File: src/main/resources/changelog/ddl/4_create_employees_table.xml [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+    http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
+
+    <changeSet id="4" author="lukas.simonelis">
+        <sql>
+            CREATE TABLE employee (
+            employee_id UUID DEFAULT RANDOM_UUID() PRIMARY KEY,
```
Comment (anton-kolyaev) [inline src/main/resources/changelog/ddl/4_create_employees_table.xml:11]: will RANDOM_UUID() work on Postgres?

File: build.gradle [hunk 15]
```diff
@@ -18,24 +18,24 @@ repositories {
...
-    annotationProcessor 'org.projectlombok:lombok'
-    implementation 'org.mapstruct:mapstruct:1.5.3.Final'
-    annotationProcessor 'org.mapstruct:mapstruct-processor:1.5.3.Final'
-    testImplementation 'org.jeasy:easy-random-core:5.0.0'
-    implementation 'org.springframework.boot:spring-boot-starter-validation'
-    runtimeOnly 'com.h2database:h2'
-  
-    // For future use
-    // implementation 'org.postgresql:postgresql:42.5.0'
+	implementation 'org.springframework.boot:spring-boot-starter'
```
Comment (anton-kolyaev) [inline build.gradle:21]: were there any reasonable updates in this file?

**Review by anton-kolyaev — CHANGES_REQUESTED** [review]

File: src/main/resources/changelog/ddl/4_create_employees_table.xml (hunk 14 above)
Comment (Cioperis) [reply src/main/resources/changelog/ddl/4_create_employees_table.xml:11]: it works with 'uuid-ossp' extension enabled + claims use random_uuid()

File: build.gradle [hunk 16]
```diff
@@ -18,30 +18,27 @@ repositories {
...
+	implementation 'org.springframework.boot:spring-boot-starter-security'
+	implementation 'org.springframework.boot:spring-boot-starter-web'
+  testImplementation 'com.h2database:h2:1.3.148'
```
[the remark this answers is not in the document]
Comment (Cioperis) [reply build.gradle]: done

File: build.gradle (hunk 15 above)
Comment (Cioperis) [reply build.gradle]: there weren't. i needed to rebuild the project in gradle menu but i tried implementing other dependencies here. unnecessary/removed.

File: src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    public List<EmployeeDTO> getAllEmployees(){
+        return employeeService.getAllEmployees();
+    }
+
+    @PutMapping("/{employeeId}")
+    public EmployeeDTO updateEmployee(@PathVariable UUID employeeId, @RequestBody EmployeeDTO updatedEmployeeDTO){
+        return employeeService.updateEmployee(employeeId, updatedEmployeeDTO);
+    }
+
+    @DeleteMapping("/{employeeId}/deactivate")
```
Comment (anton-kolyaev) [inline src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java:37]: need not in` /deactivate` suffix your final REST will be read as
```
DELETE employees {with given ID}
```

File: src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
+package com.coherentsolutions.pot.insurance.entity;
+
+import com.coherentsolutions.pot.insurance.constants.EmployeeStatus;
+import jakarta.persistence.*;
```
Comment (anton-kolyaev) [inline src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java:4]: replace wildcard with explicit imports

**Review by anton-kolyaev — CHANGES_REQUESTED** [review]

File: src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java (hunk 18 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java]: replaced

File: src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java (hunk 17 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java]: fixed

File: src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    public List<EmployeeDTO> getAllEmployees(){
+        return employeeService.getAllEmployees();
+    }
+
+    @PutMapping("/{employeeId}")
+    public EmployeeDTO updateEmployee(@PathVariable UUID employeeId, @RequestBody EmployeeDTO updatedEmployeeDTO){
+        return employeeService.updateEmployee(employeeId, updatedEmployeeDTO);
+    }
+
+    @DeleteMapping("/{employeeId}")
```
Comment (anton-kolyaev) [inline src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java:37]: HttpStatus.NO_CONTENT if you return void

File: src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @GetMapping
+    public List<EmployeeDTO> getAllEmployees(){
+        return employeeService.getAllEmployees();
+    }
+
+    @PutMapping("/{employeeId}")
+    public EmployeeDTO updateEmployee(@PathVariable UUID employeeId, @RequestBody EmployeeDTO updatedEmployeeDTO){
+        return employeeService.updateEmployee(employeeId, updatedEmployeeDTO);
```
Comment (anton-kolyaev) [inline src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java:34]: use appropriate response statuses as here
https://github.com/Java-POT-Summer-24/pot-java-summer-24/pull/19/files#diff-2a7b45711279412c3107b85210cb9da296ddf6520c76e837e68641d5d48c55ccR19

Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java:39]: done

File: src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java (hunk 19 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java:42]: void is returned so corresponding responsestatus is added

File: src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
+package com.coherentsolutions.pot.insurance.controller;
+
+import com.coherentsolutions.pot.insurance.dto.EmployeeDTO;
+import com.coherentsolutions.pot.insurance.service.EmployeeService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.*;
```
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/EmployeeController.java]: fixed wildcard

File: src/test/java/com/coherentsolutions/pot/insurance/claim/EmployeeControllerTest.java (hunk 13 above)
Comment (Cioperis) [reply src/test/java/com/coherentsolutions/pot/insurance/claim/EmployeeControllerTest.java]: done

**Review by anton-kolyaev — APPROVED** [review]

File: src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+    private String userName; // unique field
+
+    private String email;
+
+    private LocalDate dateOfBirth;
+
+    private Integer ssn;
+
+    private Integer phoneNumber;
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java:41]: I guess storing phone number as String makes work with it more flexible

File: src/main/java/com/coherentsolutions/pot/insurance/mapper/EmployeeMapper.java [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+
+@Mapper
+public interface EmployeeMapper {
+    EmployeeMapper INSTANCE = Mappers.getMapper(EmployeeMapper.class);
+
+    EmployeeDTO employeeToEmployeeDTO(EmployeeEntity employee);
+
+    EmployeeEntity employeeDTOToEmployee(EmployeeDTO employeeDTO);
+
+    void updateEmployeeFromDTO(EmployeeDTO dto, @MappingTarget EmployeeEntity entity);
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/mapper/EmployeeMapper.java:17]: ignore id?

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class EmployeeService {
+    private final EmployeeRepository employeeRepository;
+
+    public EmployeeDTO addEmployee(EmployeeDTO employeeDTO){
+        return EmployeeMapper.INSTANCE
+                .employeeToEmployeeDTO(employeeRepository.save(EmployeeMapper.INSTANCE.employeeDTOToEmployee(employeeDTO)));
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java:21]: I would suggest you keep saving-converting-converting operations separately for better readability

File: src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java (hunk 22 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java]: done

File: src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+@Entity
+@Table(name = "employee")
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class EmployeeEntity {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private UUID employeeId;
```
[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java]: done

File: src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java (hunk 24 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/service/EmployeeService.java]: unsure if i understood on how to properly do it, did some readability improvements

File: src/main/java/com/coherentsolutions/pot/insurance/mapper/EmployeeMapper.java (hunk 23 above)
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/mapper/EmployeeMapper.java:17]: fixed

File: src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (SavaLX) [reply src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java]: I meant to extend from BaseEntity which already has ID property

[the remark this answers is not in the document]
Comment (Cioperis) [reply src/main/java/com/coherentsolutions/pot/insurance/entity/EmployeeEntity.java]: fixed it now

**Review by SavaLX — APPROVED** [review]

**Review by anddevdev — APPROVED** [review]

**Pull request merged**